Improve diagnostic for throw in function type AND wrong position

alexito4 opened the pull request:
<!-- What's in this pull request? -->

This PR improves the diagnostic when `throw` is used in a function type **and** in the wrong position.

Swift currently has specific diagnostics for the following cases:

- using `throw` instead of `throws` in function type. The compiler correctly detects and suggests the proper change. 

![image (4)](https://user-images.githubusercontent.com/750912/57181462-003d3980-6e8c-11e9-8cd9-e82a9df03aa5.png)

- `throws` in the wrong position. The compiler correctly detects the wrong position and is able to move it. 

![image (4)](https://user-images.githubusercontent.com/750912/57181482-4a261f80-6e8c-11e9-84c5-d134c80cf383.png)

Both of these scenarios are really useful, but when you use `throw` in the function type and in the wrong position the error is not really useful. Furthermore executing the fixit just makes things worse. 

![image (3)](https://user-images.githubusercontent.com/750912/57181445-cb30e700-6e8b-11e9-9e40-cd39d92412e1.png)

This is not a rare mistake. I've seen it happen to newcomers to the language or even experienced programmers that just mistyped throws and moved it in the wrong place before letting the fix-it appear. 

With this changes the compiler is now able to detect the combination of both cases and give a better error. It also able to provide an appropriate fixit that will rename `throw` to `throws` and move it to the correct position. 

> expected throwing specifier; did you mean 'throws'? 
> 'throws' may only occur before '->'

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

alexito4 committed:
Improve diagnostic and fixit for throw in function type AND wrong position

theblixguy commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -724,6 +724,9 @@ ERROR(rethrows_in_wrong_position,none,
       "expected throwing specifier; did you mean 'throws'?", ())
+ERROR(throw_in_function_type_and_wrong_position,none,
+      "expected throwing specifier; did you mean 'throws'? "
+      "'throws' may only occur before '->'", ())
```
This should be in a separate note which applies a fix-it. You could then reuse the `throw_in_function_type` diagnostic. Also, it might be better to not have a separate not for this at all, but just reuse `rethrows_in_wrong_position` diagnostic. You can tweak it so it's `%select{'throws'|'rethrows'}0 may only occur before '->'` and then pass `1` when emitting this diagnostic.

theblixguy commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -724,6 +724,9 @@ ERROR(rethrows_in_wrong_position,none,
       "expected throwing specifier; did you mean 'throws'?", ())
+ERROR(throw_in_function_type_and_wrong_position,none,
+      "expected throwing specifier; did you mean 'throws'? "
+      "'throws' may only occur before '->'", ())
```
This should be in a separate note which applies a fix-it. You could then reuse the throw_in_function_type diagnostic. Also, it might be better to not have a separate not for this at all, but just reuse rethrows_in_wrong_position diagnostic. You can tweak it so it's %select{'throws'|'rethrows'}0 may only occur before '->' and then pass 1 when emitting this diagnostic.

alexito4 replied:
I would investigate this approach, reusing existing diagnostics sounds much better 👍 
I'm not sure how the notes system work yet.

alexito4 replied:
@theblixguy could you clarify what do you expect?

1. The error to only say `'throws' may only occur before '->'` but the fix it to apply both transformations?
2. The error to say `expected throwing specifier; did you mean 'throws'?` and an extra note with `'throws' may only occur before '->'`?
3. Something else? ¯\_(ツ)_/¯ 

I would like to clarify that I think it's important that the fix-it solves the 2 issues at the same time (position and renaming). That's why I think that spelling out both problems is important, but I maybe missing something. 

As a separate note, could you point me to an example where a note is attached to an error? I'm having hard time figure out how to create the `diagnose(throwsLoc, diag::throw_in_function_type)` and attach a note later, plus the fix it. 

Thanks!

theblixguy replied:
I mean 

1. Change `rethrows_in_wrong_position` to:

```
ERROR(rethrows_or_throws_in_wrong_position,none,
      "%select{'rethrows'|'throws'}0 may only occur before '->'", ())
```

2. Add a new note:

```
NOTE(throw_in_function_type_and_wrong_position_note,none,
     "'throws' may only occur before '->'", ())
``` 

3. Return `diagnose(throwsLoc, diag::throw_in_function_type_and_wrong_position, 1);`
4. Call `diagnose(throwsLoc, diag:: throw_in_function_type_and_wrong_position_note)` and then call `fixItInsert` & `fixItRemove` on it, instead of on `diagOpt`.

theblixguy commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -787,6 +788,10 @@ Parser::parseFunctionSignature(Identifier SimpleName,
+    } else if (Tok.is(tok::kw_throw)) {
+        throwsLoc = consumeToken();
+        isThrow = true;
+        return diagnose(throwsLoc, diag::throw_in_function_type_and_wrong_position);
```
Why not put all of this [here](https://github.com/apple/swift/blob/592d6603afe6646db59af57fb7cc39e21d679f21/lib/Parse/ParsePattern.cpp#L786)?

theblixguy replied:
Also, you need to indent your code with 2 spaces. You can run `git-clang-format` for this - you can run this when you're committing the file.

alexito4 replied:
Sorry I don't understand the question. The linked `if` is checking for `throws` but the new code needs to check for `throw`.

theblixguy replied:
I meant you could combine it i.e “if token == throw OR token == throws” and then consume the token and set isThrow to true.

Then, check if you set isThrow=true and return the diagnostic. 

Anyway, it was just an optional suggestion :)

alexito4 committed:
Run clang-format

alexito4 committed:
Change diagnostic to use an existing error + new note with fixits

alexito4 commented:
Thanks for the feedback and the help!

I've implemented some changes following the suggestions.

- I'm now using a NOTE `throw_in_function_type_and_wrong_position_note` and reusing the existing error `throws_in_wrong_position`. The tests have been updated accordingly.
- To attach a note I had to `flush` the existing diagnostic, otherwise I got `Already have an active diagnostic` error. This seems to be done in other places but please confirm if it's correct.
- I haven't implemented the suggestion to unify "throws" and "rethrows" in wrong place errors. It doesn't seem required for this change as those errors are already used currently. Although it seems like a nice little refactor to improve this part of the code, so I tried to do it and a bunch of compiler errors started popping up 🙃 which I fixed successfully until I got into a template error 😅. 
- The code I'm not happy with is the one in l.855. Obviously it should reuse the fixits by just applying it into a variable that holds the original diagnostic or the new note. But everything I tried to assign the `diagOpt` to another variable failed with template errors in the Optional type. Seems like my c++ is too rusty, suggestions welcomed. 

As an extra comment, I still think the previous diagnostic was a little better in one aspect, the order. Right now we mention `throws' may only occur before '->'` and then `expected throwing specifier; did you mean 'throws'?`, seems weird to me that we talk about `throws` before it's been suggested.

CodaFi commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -724,6 +724,8 @@ ERROR(rethrows_in_wrong_position,none,
 ERROR(throw_in_function_type,none,
       "expected throwing specifier; did you mean 'throws'?", ())
+NOTE(throw_in_function_type_and_wrong_position_note,none,
+      "expected throwing specifier; did you mean 'throws'?", ())
```
Let’s not surface the word “specifier” to users here.  You can remove the first clause before the semicolon and just use the dependent clause for the fixit.

CodaFi replied:
I think we can be more specific.  The dependent clause is good, but the clause before the semicolon doesn’t actually reflect the mistake the user made.

alexito4 replied:
What do you prefer instead then?

The user has made two mistakes (wrong keyword and in wrong position) and I think both should be surfaced.

Right now 1 is displayed in the error and the other in the note. 
In my original commit the error displayed both at the same time.

xwu replied:
I would expect one error message only to explain the problem. Something like, "unexpected 'throw' in function type; did you mean 'throws', which may occur only before '->'?"

Agree that "specifier" isn't a good user-facing word and the existing diagnostic could use improvement too.

alexito4 replied:
> I would expect one error message only to explain the problem.

That was my initial expectation also :) In case we want just one error I would rollback to 592d660 and apply the copy changes, which look like an improvement anyway 👍 

If somebody can confirm which behaviour we prefer (1 errors vs. error + note) I can continue with this.

CodaFi commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -830,15 +843,25 @@ Parser::parseFunctionSignature(Identifier SimpleName,
-              .fixItRemoveChars(typeEndLoc, throwsEndLoc);
+
+    if (isThrow) {
+      (*diagOpt).flush();
```
Any particular reason you need the extra flush?

alexito4 replied:
Yes, without it got "Already have an active diagnostic" error. My understanding was that because i'm doing a diagnose of an error and then another diagnose of the note they were conflicting so I had to flush he first one. I found it done this way in other parts of the code.

CodaFi commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -807,13 +812,21 @@ Parser::parseFunctionSignature(Identifier SimpleName,
-      (*diagOpt).fixItExchange(SourceRange(arrowLoc),
-                               SourceRange(throwsLoc));
+      if (isThrow) {
+        (*diagOpt).flush();
```
Same here.  Extra flush?

futurejones commented:
Can one of the admins verify this patch?

alexito4 commented:
I'm closing this since it didn't go anywhere and I would have a hard time picking it back and updating it to master 😅
